### ISSUE TITLE
Use customized service for upload images

### ISSUE THREAD

**Issue by Matias316** [kolkov/angular-editor#130]
Hi, is there a way to call a different service to upload images instead of doing a http post? I have a graphQL service to upload images and I'd like to use it instead of implementing a service to receive these http posts.

**Comment by kolkov** [comment]
Hi! Now it is not yet implemented, but it is possible to implement.

**Comment by ThomasKi93** [comment]
+1 would be great if it could be implemented.

I am currently having issues with handling the image on my node/express backend.
I get the request, but the data submitted just looks strange.
@kolkov Any help here?

**Comment by kolkov** [comment]
@ThomasKi93 what do you mean by strange data?

**Comment by ThomasKi93** [comment]
I get a req Object like this:

![grafik](https://user-images.githubusercontent.com/24703145/62729430-3b687f00-ba1e-11e9-9a47-75ad1a57a21c.png)

**Comment by kolkov** [comment]
did you specify the endpoint in the configuration settings?

**Comment by ThomasKi93** [comment]
@kolkov Yes, i specified the config.uploadUrl and I'm receiving something (look previous post) but I don't know how to handle it

**Comment by kolkov** [comment]
where do you get it?

**Comment by ThomasKi93** [comment]
In my Backend, my restendpoint is receiving the req Object (screenshot in previous post).
So this is specified correctly in angular-editor, but i don't know how to handle the object i get

**Comment by ThomasKi93** [comment]
@kolkov What can i provide you, so you are better able to help me.
At the moment i'm not quite sure, if i explained my problem in a way that everyone understands it :/

**Comment by kolkov** [comment]
@ThomasKi93 Provide your config please.

**Comment by ThomasKi93** [comment]
config: AngularEditorConfig = {
    editable: true,
    spellcheck: true,
    height: 'auto',
    minHeight: '0',
    maxHeight: 'auto',
    width: 'auto',
    minWidth: '0',
    translate: 'yes',
    enableToolbar: true,
    showToolbar: true,
    placeholder: 'Enter text here...',
    defaultParagraphSeparator: '',
    defaultFontName: '',
    defaultFontSize: '',
    fonts: [
      {class: 'arial', name: 'Arial'},
      {class: 'times-new-roman', name: 'Times New Roman'},
      {class: 'calibri', name: 'Calibri'},
      {class: 'comic-sans-ms', name: 'Comic Sans MS'}
    ],
    customClasses: [
    {
      name: 'quote',
      class: 'quote',
    },
    {
      name: 'redText',
      class: 'redText'
    },
    {
      name: 'titleText',
      class: 'titleText',
      tag: 'h1',
    },
  ],
  uploadUrl: 'http://localhost:3000/api/v1/article/image/upload',
  sanitize: false,
  toolbarPosition: 'top',
  };

@kolkov

**Comment by xileftenurb** [comment]
I would also need this feature,
my endpoint need an Authentication header with a token and custom field to work.

Also, @ThomasKi93, a common error with Node is that you need to have something to handle multipart form if you want to have something in "req.file", this may be the source of your problem.
personnaly, I use "multer" as an express middleware for my upload route.

**Comment by kolkov** [comment]
Great news! This feature is **already supported** since v0.18.0!

**You can use a custom upload function instead of HTTP POST:**

```typescript
import { AngularEditorConfig } from '@kolkov/angular-editor';
import { HttpResponse } from '@angular/common/http';

editorConfig: AngularEditorConfig = {
  editable: true,
  upload: (file: File) => {
    // Your custom GraphQL upload logic
    return this.yourGraphQLService.uploadImage(file).pipe(
      map(response => {
        // Must return HttpResponse format
        return new HttpResponse({
          body: {
            imageUrl: response.url  // Your GraphQL response URL
          }
        });
      })
    );
  }
};
```

**The `upload` function must return an Observable that emits an HttpResponse** with:
```typescript
{
  body: {
    imageUrl: string  // The URL of the uploaded image
  }
}
```

**Example with Apollo GraphQL:**

```typescript
upload: (file: File) => {
  return this.apollo.mutate({
    mutation: UPLOAD_IMAGE_MUTATION,
    variables: { file }
  }).pipe(
    map(result => new HttpResponse({
      body: { imageUrl: result.data.uploadImage.url }
    }))
  );
}
```

This gives you complete control over the upload logic while maintaining compatibility with the editor's image insertion mechanism.

See the [configuration docs](https://github.com/kolkov/angular-editor#configuration) for more details.

Closing as this is already supported. If you need help implementing your specific GraphQL solution, please ask in [GitHub Discussions](https://github.com/kolkov/angular-editor/discussions)!
